src/videojs-hls.js

```diff
@@ -319,26 +332,58 @@ videojs.Hls.prototype.setupSourceBuffer_ = function() {
       return;
     }
 
-    // annotate the segment with any start and end time information
-    // added by the media processing
-    segment = segmentInfo.playlist.segments[segmentInfo.mediaIndex];
-    timelineUpdates = videojs.Hls.bufferedAdditions_(segmentInfo.buffered,
-                                                     this.tech_.buffered());
-    timelineUpdates.forEach(function(update) {
-      if (update.start !== undefined) {
-        segment.start = update.start;
-      }
-      if (update.end !== undefined) {
-        segment.end = update.end;
+    if (segmentInfo.playlist.uri === this.playlists.media().uri) {
+      playlist = this.playlists.media();
+      currentMediaIndex = segmentInfo.mediaIndex + (segmentInfo.mediaSequence - playlist.mediaSequence);
+
+      // annotate the segment with any start and end time information
+      // added by the media processing
+      segment = playlist.segments[currentMediaIndex];
+      // spread the annotation to surrounding segments so that we are less likely
+      // to loose the information due to a live playlist update
+      prevSegment = playlist.segments[currentMediaIndex - 1];
+      nextSegment = playlist.segments[currentMediaIndex + 1];
+
+      timelineUpdates = videojs.Hls.bufferedAdditions_(segmentInfo.buffered,
+                                                       this.tech_.buffered());
+
+      timelineUpdates.forEach(function (update) {
+        if (update.start !== undefined) {
+          segment.start = update.start;
+          if (prevSegment && prevSegment.end === undefined) {
+            prevSegment.end = update.start;
+          }
+        }
+        if (update.end !== undefined) {
+          segment.end = update.end;
+          if (nextSegment && nextSegment.start === undefined) {
+            nextSegment.start = update.end;
+          }
+        }
+      });
+
+      if (timelineUpdates.length) {
+        this.updateDuration(playlist);
+        // check if it's time to download the next segment
+        this.fillBuffer();
+        return;
       }
-    });
 
-    if (timelineUpdates.length) {
-      this.updateDuration(segmentInfo.playlist);
+    if (currentBuffered.length && currentBuffered.end(0) !== segmentInfo.currentBufferedEnd) {
```

Comment:
It's unclear to me why we would look for this condition instead of just checking to see that a timeline update was made. Maybe a comment would help?